File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.py
def solution(n, costs):
    graph = {}
    
    for start, end, cost in costs:
        if start not in graph:
            graph[start] = []
        if end not in graph:
            graph[end] = []
        graph[start].append((end, cost))
        graph[end].append((start, cost))
    
    total_cost = 0
    visited = set()  
    visited.add(0)   
    
    while len(visited) < n:
        min_cost = float('inf')
        next_node = None
        
        for node in visited:
            for neighbor, cost in graph[node]:
                if neighbor not in visited and cost < min_cost:
                    min_cost = cost
                    next_node = neighbor
        
        visited.add(next_node)
        total_cost += min_cost
    
    return total_cost

File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] costs) {
        List<List<int[]>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }

        for (int[] edge : costs) {
            int from = edge[0];
            int to = edge[1];
            int cost = edge[2];
            graph.get(from).add(new int[]{to, cost});
            graph.get(to).add(new int[]{from, cost});
        }

        boolean[] visited = new boolean[n];
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[1]));
        int totalCost = 0;

        visited[0] = true;
        for (int[] neighbor : graph.get(0)) {
            pq.offer(neighbor);
        }

        while (!pq.isEmpty()) {
            int[] current = pq.poll();
            int to = current[0];
            int cost = current[1];

            if (visited[to]) continue;

            visited[to] = true;
            totalCost += cost;

            for (int[] neighbor : graph.get(to)) {
                if (!visited[neighbor[0]]) {
                    pq.offer(neighbor);
                }
            }
        }

        return totalCost;
    }
}
